### DISCUSSION TITLE
sub stuck

### DISCUSSION THREAD

**Discussion by LePiromano** [FakeFishGames/Barotrauma#15054] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

the issue where the submarine gets stuck in level geometry forcing a softlock is still present, i'm sure it has been reported before but i couldn't find an open ticket
ramming into a tight hole can OFTEN get you stuck, specially when it's inbetween a ruin and a level wall for some reason.
once you get stuck you can't create enough speed to get out, maybe if the submarine is stuck trying to move at maximum velocity but not changing position for a set amount of time the game could detect it? 

### Reproduction steps

find a tight hole in the level inbetween a wall and a wreck/alien ruin
make sure you're using a submarine that has any slight protrusions that could get stuck (eg humpback)
ram into the hole full speed
can't get out

### Bug prevalence

Happens every now and then

### Single player or multiplayer?

Single player

### Version

v1.7.0.1 (Unstable)

### Which operating system did you encounter this bug on?

Windows

**Comment by SomeRandomNoobKekeke** [comment]
https://github.com/FakeFishGames/Barotrauma/issues/4632
Yep, tried hard and was able to get stuck on humpback in
```
Level seed: nV7KKqtY
Level generation params: wastesbasic
```
(Also level generated by editor with same seed mismatch quickstart level, should i bug report it? xd)
![20241104023510_1](https://github.com/user-attachments/assets/1845ca0c-e0aa-40cc-8a37-2fb4a6a7bb53)
I think it's wide known problem, but no one reports it because:
> ram into the hole full speed
can't get out

What did you expect? A head pats?

And it's smells like problem with general sloppiness of physics engine
How would you fix "general sloppiness of physics engine"?
E.g. what if sub isn't stuck but trapped in some pocket and can move a little but can't get out?
How would you detect that?
E.g. in l4d2 it was solved by manually blocking stuck points with invisible colliders
Idk how to do this in procedurally generated caves
![barodev small](https://github.com/user-attachments/assets/7c00c095-bc75-408d-83ad-6c3c10e319e6)

**Comment by MKMoose** [comment]
A simple "Unstuck" button or command (as suggested on Discord by Crazyy) is all you need, really, since the issue isn't common enought to warrant changes to level generation or to the physics engine.

Probably just make it teleport to the closest path or abyss node if possible, otherwise to the closest main path node. Or get a little more fancy if you feel like it to find a more suitable target.

Suggested conditions (not necessary, especially if it's only a command):
- host only,
- disallow it if the sub's moving or has been moving recently,
- disallow it while there are hostiles nearby (ideally also account for PvP, even between members of the same team, might want to check for any nearby characters firing a weaopon/turret or receiving an attack or sth),
- the sub has to be in good shape (especially no broken engines, ballast pumps, reactors and so on, though ignore noninteractable ones),
- if needed for some rare cases, ignore some or all of the above (primarily the hostiles one) after the sub hasn't been moving or has moved very little over an extended period of time,
- cooldown to avoid spamming.

**Reply by SomeRandomNoobKekeke** [reply]
Done https://steamcommunity.com/sharedfiles/filedetails/?id=3360208492

it checks sub worldposition constantly and adds it to queue of 10 if sub traveled a least 1m from previous position
And "unstuck" command then teleports sub 10 steps back and flushes the queue
And i don't care about cheats or permissions because it's just 10m
Also there's 20sec cooldown to prevent spam to stay in place, e.g. when falling to the abyss
